### ISSUE TITLE
Request review of OCP FairCoin code

### ISSUE THREAD

**Issue by bhaugen** [FreedomCoop/valuenetwork#10]
## Wrapping FairCoin objects in OCP objects

* [A FairCoin address is wrapped in an OCP EconomicResource](https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/models.py#L3934).
* EconomicAgents (such as Freedom Coop workers) may own these [FairCoin address resources](https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/models.py#L462) in the single FairCoin wallet used on the OCP server.
* [FairCoin transactions are wrapped in OCP EconomicEvents](https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/models.py#L10066).

## FairCoin transactions do not get sent directly from Django

* I tried that a couple of ways. Both failed, with symptoms that said it would be difficult or impossible.
* Only [the wrapper EconomicEvents get created with the necessary FairCoin information](https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/views.py#L1323).

## A cron job creates and broadcasts the FairCoin transactions to the network.

* [The command](https://github.com/FreedomCoop/valuenetwork/blob/master/work/management/commands/broadcast_faircoins.py) to be put on a crontab.
* [The code that broadcasts the transactions](https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/faircoin_utils.py#L75).
* [The lock around the cron job](https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/lockfile.py) in case it is not finished when the crontab starts it again.

### Every time the wrapper EconomicEvent is used for anything, we check the transaction state

(until it is confirmed) 

* [Transaction state update code](https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/models.py#L10272)

**Comment by bhaugen** [comment]
Current state: the broadcast_faircoins command works fine when called from the command line, but does not do anything when called from a cronjob. syslog says the cronjob is running every minute, but also says "info (No MTA installed, discarding output)".
Here's my crontab: 
```
* * * * * (cd /home/bob/.virtualenvs/fcx/valuenetwork; python manage.py broadcast_faircoins  >> /home/bob/.virtualenvs/fcx/valuenetwork/broadcast_faircoins.log)
```
I thought redirecting to a file would result in some output in the file. Apparently not, and I am using both logging and print statements.

Any ideas?

**Comment by bhaugen** [comment]
More clues: first, my crontab command was not actually executing, but I could not tell because I couldn't get logging to go to a file. So I needed to change the crontab to this:
```
* * * * * /home/bob/.virtualenvs/fcx/bin/python /home/bob/.virtualenvs/fcx/valuenetwork/manage.py broadcast_faircoins >> /home/bob/.virtualenvs/fcx/valuenetwork/broadcast_faircoins.log  2>&1
```
The full qualification of python and manage.py was important for running from cron, and so was the ```2>&1``` at the end.

Then I got this error message:
```
Traceback (most recent call last):
  File "/home/bob/.virtualenvs/fcx/valuenetwork/manage.py", line 9, in <module>
    execute_from_command_line(sys.argv)
  File "/home/bob/.virtualenvs/fcx/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 443, in execute_from_command_line
    utility.execute()
  File "/home/bob/.virtualenvs/fcx/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 382, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/bob/.virtualenvs/fcx/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 261, in fetch_command
    klass = load_command_class(app_name, subcommand)
  File "/home/bob/.virtualenvs/fcx/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 69, in load_command_class
    module = import_module('%s.management.commands.%s' % (app_name, name))
  File "/home/bob/.virtualenvs/fcx/local/lib/python2.7/site-packages/django/utils/importlib.py", line 35, in import_module
    __import__(name)
  File "/home/bob/.virtualenvs/fcx/valuenetwork/work/management/commands/broadcast_faircoins.py", line 16, in <module>
    from valuenetwork.valueaccounting.faircoin_utils import broadcast_tx
  File "/home/bob/.virtualenvs/fcx/valuenetwork/valuenetwork/valueaccounting/faircoin_utils.py", line 5, in <module>
    import faircoin_nrp.electrum_fair_nrp as efn
  File "/home/bob/.virtualenvs/fcx/valuenetwork/faircoin_nrp/electrum_fair_nrp.py", line 35, in <module>
    wallet_path = config.get('electrum','wallet_path')
  File "/usr/lib/python2.7/ConfigParser.py", line 607, in get
    raise NoSectionError(section)
ConfigParser.NoSectionError: No section: 'electrum'
```
Which I suspect is also about the cronjob not finding the config file...

**Comment by bhaugen** [comment]
Yes, that error message was fixed by adding the full path to ```config.read("/home/bob/.virtualenvs/fcx/valuenetwork/electrum-fair-nrp.conf")```.

But I only get my first logger message in the output file, because I think electrum_fair_nrp.py reconfigures logging...

**Comment by bhaugen** [comment]
That's my current stuck point. The cronjob is running, but not doing its faircoin transaction work, and I am not getting any more error messages, so I don't know why. The errors are being swallowed somewhere in the mix.

**Comment by bhaugen** [comment]
By the way, the >> /home/bob/.virtualenvs/fcx/valuenetwork/broadcast_faircoins.log was not needed, and was not actually doing anything. I also had the same file in a filehandler connected to the logger, and did not think that was working, so added a belt and suspenders. That was before I added ```2>&1```.
All I needed was ```* * * * * /home/bob/.virtualenvs/fcx/bin/python /home/bob/.virtualenvs/fcx/valuenetwork/manage.py broadcast_faircoins  2>&1```.

But I am still not getting any other logging messages after the first one. All the messages about the actual faircoin tx are being swallowed.

**Comment by bhaugen** [comment]
Here's where I am at now. Still stuck.

Here's the [point in the code](https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/faircoin_utils.py#L90). Start there and look for two more instances of the word problem.

Summary:
* Calling efn.make_transaction_from_address() directly from Django did not work because Django just hung.
* Calling my management command [broadcast_faircoins](https://github.com/FreedomCoop/valuenetwork/blob/master/work/management/commands/broadcast_faircoins.py) works from the command line *sometimes*, but does not work from a cronjob.
    * Even from the command line, sometimes the same address_origin, address_end and amount will fail, and then if I run the command again, the same input will succeed in broadcasting a FC transaction.
    * From the command line, I get all of the logging messages written to the designated filehandler file.
    * But from cron, the logging messages stop, and the faircoin transaction never gets created, and I have no clues about why.

I'm using a cron job here instead of another daemon because I have some limited experience with those, although obviously not enough (I have no idea what's going on with the missing logging yet). But I have no experience with daemons.

Short version: I in over my head, and trying to keep swimming. It might help to see the actual code FairMarket is using to make faircoin transactions.

**Comment by bhaugen** [comment]
Breakthrough, finally! Got the cronjob to handle some transactions successfully!

Not totally out of the woods yet, but I can see daylight.

**Comment by bhaugen** [comment]
Cron job is now totally working, including the lockfile. Logging seems to be under control.

Only got one more test to run before I declare victory on the code, until we get all this stuff deployed and then of course things will break in production...

**Comment by bhaugen** [comment]
Oops, stalled again. Current blocker is that if the cronjob is running, creating new faircoin addresses from the django UI hangs. I tried creating another module so that elecrum_fair_nrp.py was running in the cronjob and the other module was creating new  faircoin addresses, but that did not help.
